ipa-metrics/src/label.rs

```diff
@@ -0,0 +1,172 @@
+use std::{
+    fmt::{Debug, Display, Formatter},
+    hash::{Hash, Hasher},
+};
+
+pub const MAX_LABELS: usize = 5;
+
+/// Dimension value (or label value) must be sendable to another thread
+/// and there must be a way to show it
+pub trait LabelValue: Display + Send {
+    /// Creates a unique hash for this value.
+    /// It is easy to create collisions, so better avoid them,
+    /// by assigning a unique integer to each value
+    ///
+    /// Note that this value is used for uniqueness check inside
+    /// metric stores
+    fn hash(&self) -> u64;
+
+    /// Creates an owned copy of this value. Dynamic dispatch
+    /// is required, because values are stored in a generic store
+    /// that can't be specialized for value types.
+    fn boxed(&self) -> Box<dyn LabelValue>;
+}
+
+impl LabelValue for u32 {
+    fn hash(&self) -> u64 {
+        u64::from(*self)
+    }
+
+    fn boxed(&self) -> Box<dyn LabelValue> {
+        Box::new(*self)
+    }
+}
+
+pub struct Label<'lv> {
+    pub name: &'static str,
+    pub val: &'lv dyn LabelValue,
+}
+
+impl Label<'_> {
+    #[must_use]
+    pub fn to_owned(&self) -> OwnedLabel {
+        OwnedLabel {
+            name: self.name,
+            val: self.val.boxed(),
+        }
+    }
+}
+
+impl Debug for Label<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Label")
+            .field("name", &self.name)
+            .field("val", &format!("{}", self.val))
+            .finish()
+    }
+}
+
+impl Hash for Label<'_> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write(self.name.as_bytes());
```

Comment:
Does it make sense to change this to `Hash::hash`? I think that makes this prefix free, which in turn makes the name hashing prefix free.